Use smart pointers in WebXROpaqueFramebuffer

mwyrzykowski opened the pull request:
#### e3d6e0875a7a26b3fe200084914de415ce5dc1c7
<pre>
Use smart pointers in WebXROpaqueFramebuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=285629">https://bugs.webkit.org/show_bug.cgi?id=285629</a>
<a href="https://rdar.apple.com/122475331">rdar://122475331</a>

Reviewed by Dan Glastonbury.

WebXROpaqueFramebuffer stored a C++ reference to the WebGL context,
but it is not clear who ensures the GL context is retained during
the lifetime of the WebXROpaqueFramebuffer.

Via code inspection, the out of bounds access should only occur
if the reference was invalid or, more likely, WebGLRenderingContextBase::initializeContextState()
was never called, in which case the store / resolve should be a nop.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
Bail if the WeakPtr is null.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
Store a WeakPtr instead of a &amp;.

* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedWebGLRestoreTexture::~ScopedWebGLRestoreTexture):
Ensure we don&apos;t access out of bounds in the texture units.

Canonical link: <a href="https://commits.webkit.org/291611@main">https://commits.webkit.org/291611@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a0cc04f82d9b6d9877b78be9f780dbe522a8f88

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98445 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43970 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21460 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71385 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28767 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9945 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84497 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51719 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9627 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2126 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43284 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2155 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100476 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20496 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80400 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20748 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80420 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79729 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19821 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24266 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13622 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20480 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25658 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23627 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -78,76 +82,84 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
-        blitSharedToLayered(*gl);
-        break;
+    RefPtr refContext = m_context.get();
+    if (refContext) {
```
Can we early out here instead of reformatting the entire function?

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -78,76 +82,84 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
-    gl->bindFramebuffer(GL::FRAMEBUFFER, m_drawFramebuffer->object());
-    gl->clear(clearMask);
+    RefPtr refContext = m_context.get();
+    if (refContext) {
```
Can we early out here instead of reformatting the entire function?

djg commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -231,6 +231,9 @@ class ScopedWebGLRestoreTexture {
     ~ScopedWebGLRestoreTexture()
     {
+        if (m_context->m_textureUnits.size() <= m_context->m_activeTextureUnit)
+            return;
```
Is this to guard against `m_context` becoming invalid? Maybe a check of `m_context` would be more appropriate?

mwyrzykowski replied:
Yes, specifically `m_context->m_context` i.e., `m_context->graphicsContextGL()`

mwyrzykowski replied:
That is the only way I can see how `m_activeTextureUnit` would exceed `m_textureUnits.size()` via inspection, unless I am failing to see some code path

djg reviewed (approved):
Approved, conditional on nits. :shipit:

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -53,19 +53,23 @@ std::unique_ptr<WebXROpaqueFramebuffer> WebXROpaqueFramebuffer::create(PlatformX
     : m_handle(handle)
     , m_drawFramebuffer(WTFMove(framebuffer))
-    , m_context(context)
+    , m_context(&context)
```
`&` is not needed

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -53,19 +53,23 @@ std::unique_ptr<WebXROpaqueFramebuffer> WebXROpaqueFramebuffer::create(PlatformX
         m_resolveAttachments.release(*gl);
         m_resolvedFBO.release(*gl);
-        m_context.deleteFramebuffer(m_drawFramebuffer.ptr());
+        context->deleteFramebuffer(m_drawFramebuffer.ptr());
```
Don't we need a RefPtr / Ref to m_drawFramebuffer on the stack before passing it to deleteFramebuffer()?

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -78,7 +82,12 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
+    if (!refContext)
+        return;
+
+    auto& context = *refContext;
```
This looks weird. Why not rename `refContext` to `context` above and use `context->` below?

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -122,7 +131,12 @@ void WebXROpaqueFramebuffer::startFrame(PlatformXR::FrameData::LayerData& data)
+    if (!refContext)
+        return;
+
+    auto& context = *refContext;
```
Same weird pattern as above. We don't do protector variables anymore. You should be dealing with raw pointers directly as much as possible.

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -132,7 +132,7 @@ class WebXROpaqueFramebuffer {
     PlatformXR::LayerHandle m_handle;
     Ref<WebGLFramebuffer> m_drawFramebuffer;
-    WebGLRenderingContextBase& m_context;
+    WeakPtr<WebGLRenderingContextBase> m_context;
```
Can this really become null? It used to be a C++ reference so the change to a `WeakPtr` is surprising. SHouldn't this be a `WeakRef`?

mwyrzykowski replied:
WeakRef sounds better

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -114,12 +114,17 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
+    if (!refContext)
+        return;
+
+    auto& context = *refContext;
```
Same Weird pattern.

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -114,12 +114,17 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
     releaseAllDisplayAttachments();
 
-    if (RefPtr gl = m_context.graphicsContextGL()) {
+    RefPtr refContext = m_context.get();
```
This is basically the same as `protectedContext` variables like we used to have and that we now discourage.

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -132,7 +137,12 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
+    if (!refContext)
+        return;
+
+    auto& context = *refContext;
```
Same weird pattern.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -231,6 +231,9 @@ class ScopedWebGLRestoreTexture {
 
     ~ScopedWebGLRestoreTexture()
     {
+        if (!m_context->graphicsContextGL() || m_context->m_textureUnits.size() <= m_context->m_activeTextureUnit)
```
How does this relate to smart pointer adoption?

kkinnunen-apple reviewed (changes requested):
> WebXROpaqueFramebuffer stored a C++ reference to the WebGL context,
but it is not clear who ensures the GL context is retained during
the lifetime of the WebXROpaqueFramebuffer.

This is not correct. It may not be clear to you, but it's clear and well defined in general.
WebGLRenderingContextBase owns the context.

> Via code inspection, the out of bounds access should only occur
if the reference was invalid or, more likely, WebGLRenderingContextBase::initializeContextState()
was never called, in which case the store / resolve should be a nop.

In general, this code inspection is not correct.   WebGLRenderingContextBase::initializeContextState() is always called.

It is essential that you post fixes after you understand the code. We should not post modifications without understanding.

It is typically required to have a test case to to understand the problem and prove that the problem went away with the fix.

The work on WebXR should involve investment in the infrastucture. This means to fix WebXR bug, we should account for implementing whatever is currently missing from being able to do proper engineering. This means, the cost of fixing bugs should include the cost of doing whatever needs to be done to be able to run tests.


As for the bug itself, the bug is likely due to something related to followinlg:
- The start frame callback being scheduled 
- WebGL context is collected
- The start frame callback is run

If so, this  doesn't make sense:
A) why would WebGL context be collected if somebody is producing frames to it?
B) why would somebody produce frames if WebGL context is collectable, e.g. cannot affect anything?

So WebXR implementation is holding references incorrectly. Either A or B is in conflict with reality

This class of problems have been implemented for the base WebGL, since it has the extension objects and non-context objects.  It seems WebXR was implemented incorrectly, apparently not using the proper mechanisms. Same mechanisms should be used for WebXR object graphs, *for all those WebXR objects (layesr, opaque framebuffer, ...)*.  If the existing WebGL mechanism is not appropriate, we should identify the issue and fix that. 


Alternative source of the bug is when WebGL context is lost. For us currently, this may happen only when GPUP crashes. These class of fixes involve checking that the WebGLRenderingContextBase   is not lost -- there's a well-understood pattern for this also in the existing WebGLRenderingContext and extension code. Again: this, too, should be testable from WebGL perspective, so the logic should be usable in WebXR tests written for the fix.

mwyrzykowski committed:
Use smart pointers in WebXROpaqueFramebuffer
https://bugs.webkit.org/show_bug.cgi?id=285629
rdar://122475331

Reviewed by Dan Glastonbury.

WebXROpaqueFramebuffer stored a C++ reference to the WebGL context,
but it is not clear who ensures the GL context is retained during
the lifetime of the WebXROpaqueFramebuffer.

Via code inspection, the out of bounds access should only occur
if the reference was invalid or, more likely, WebGLRenderingContextBase::initializeContextState()
was never called, in which case the store / resolve should be a nop.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
Bail if the WeakPtr is null.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
Store a WeakPtr instead of a &.

* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedWebGLRestoreTexture::~ScopedWebGLRestoreTexture):
Ensure we don't access out of bounds in the texture units.

Canonical link: https://commits.webkit.org/291611@main
